src/core/control/jobs/RenderJob.cpp

```diff
@@ -79,12 +80,28 @@ void RenderJob::run() {
 
         std::lock_guard lock(this->view->drawingMutex);
         std::swap(this->view->crBuffer, newBuffer);
+        repaintPage();
     } else {
-        for (Rectangle<double> const& rect: rerenderRects) { rerenderRectangle(rect); }
+        for (Rectangle<double> const& rect: rerenderRects) {
+            rerenderRectangle(rect);
+            repaintPageArea(rect.x, rect.y, rect.x + rect.width, rect.y + rect.height);
+        }
     }
+}
+
+void RenderJob::repaintPage() {
+    repaintWidgetArea(view->xournal->getWidget(), 0, 0, view->getWidth(), view->getHeight());
+}
+
+void RenderJob::repaintPageArea(double x1, double y1, double x2, double y2) {
+    double zoom = view->xournal->getZoom();
+    int x = view->getX();
+    int y = view->getY();
+    repaintWidgetArea(view->xournal->getWidget(), x + std::floor(zoom * x1), y + std::floor(zoom * y1), x + std::ceil(zoom * x2), y + std::ceil(zoom * y2));
+}
 
-    // Schedule a repaint of the widget
-    repaintWidget(this->view->getXournal()->getWidget());
+void RenderJob::repaintWidgetArea(GtkWidget* widget, int x1, int y1, int x2, int y2) {
```

Comment:
This is just a helper function. No need to declare it in the header file. Just make it like so (maybe move it a little up to satisfy the compiler).

```suggestion
static void repaintWidgetArea(GtkWidget* widget, int x1, int y1, int x2, int y2) {
```